torch/_inductor/comm_analysis.py

```diff
@@ -0,0 +1,179 @@
+import math
+
+from .utils import get_dtype_size
+from . import ir
+from .utils import sympy_product
+from .virtualized import V
+
+ncclFuncAllReduce = "allreduce"
+ncclFuncAllGather = "allgather"
+ncclFuncReduceScatter = "reducescatter"
+
+NCCL_HW_NVLINK = 0
+NCCL_HW_PCI = 1
+NCCL_HW_NET = 2
+
+NCCL_ALGO_TREE = 0
+NCCL_ALGO_RING = 1
+
+# Latencies in us
+baseLat = [
+    6.8,  # Tree (LL)
+    6.6,  # Ring (LL)
+]
+# Latencies in us
+hwLat = [
+    # NVLINK
+    [
+        .6,  # Tree (LL)
+        .6,  # Ring (LL)
+    ],
+    # PCI
+    [
+        1.0,  # Tree (LL)
+        1.0,  # Ring (LL)
+    ],
+    # NET
+    [
+        5.0,  # Tree (LL)
+        2.7,  # Ring (LL)
+    ],
+]
+
+VOLTA_COMPCAP_IDX = 0
+AMPERE_COMPCAP_IDX = 1
+HOPPER_COMPCAP_IDX = 2
+
+# LL128 max BW per channel
+llMaxBws = [
+    # Volta-N1/Intel-N2/Intel-N4
+    [39.0, 39.0, 20.4],
+    # Ampere-N1/AMD-N2/AMD-N4
+    [
+        87.7,
+        22.5,  # avg of ring & tree
+        19.0,
+    ],
+    # Hopper-N1/AMD-N2/AMD-N4
+    [
+        87.7,
+        22.5,  # avg of ring & tree
+        19.0,
+    ],
+]
+
+bwNVLINK = 300  # unit: GB/s, uni-directional P2P bandwidth per card
+bwInfiniBand = 25  # unit: GB/s, uni-directional P2P bandwidth per node
+
+
+def get_collective_type(snode: "scheduler.BaseSchedulerNode") -> str:
+    if isinstance(snode.node, (ir.AllReduce, ir.AllReduceCoalesced)):
+        return ncclFuncAllReduce
+    elif isinstance(snode.node, (ir.AllGatherIntoTensor, ir.AllGatherIntoTensorCoalesced)):
+        return ncclFuncAllGather
+    elif isinstance(snode.node, (ir.ReduceScatterTensor, ir.ReduceScatterTensorCoalesced)):
+        return ncclFuncReduceScatter
+    else:
+        raise Exception(f"Unsupported collective type: {snode.node}")
+
+
+def estimate_nccl_collective_runtime(snode: "scheduler.BaseSchedulerNode") -> float:
```

Comment:
Tested it for allreduce on 2-GPU:

actual vs. predicted
[4,4], 2-gpu: 7us vs 7.8us
[16,16], 2-gpu: 7us vs 7.8us
[128, 128], 2-gpu: 8.5us vs. 10us
[1024, 1024], 2-gpu: 300us vs 50us
[2048, 2048], 2-gpu: 284us vs 186us
[10240, 10240], 2-gpu: 6146us vs 4462us

it's not too bad (maybe like +/- 30% accuracy)

Reply:
well.. more than 2 gpu is also important, i'm guessing the variance only goes higher in larger scale.

more important than how accurate the model is, how sensitive is the optimization to wrong model output?

Reply:
upon looking more closely, i am not in favor of landing this cost model.  It seems too specialized on the small-scale and currently used hw setup.  IF its important to have this much detail in a model, that implies its important for it to be accurate at large scale and across different system types.  IF its not important for this to be accurate, just replace it with a simpler heuristic.

Reply:
I believe accurate is still the best, but it's not super important currently because a somewhat (not 100%) accurate model is still better than a dummy model.

I'd actually argue that a simpler heuristic will not be much simpler in terms of the code, because we still need to consider both latency and bandwidth terms, and the computation will be quite similar to current code. E.g. for bandwidth calculation which is important for large-scale case, there is a fixed formula for calculating steps used in ring algorithm, and we will need to use the same formula in simpler heuristics. And for small-scale, latency dominates and we don't know the empirical latency numbers unless we copy from existing NCCL code.

To make this more generic, I can add V100 and H100 heuristics to the code. I believe NVLink for intra-node is mostly safe to assume, and I can add a user-facing config for intra-node interconnect bandwidth (e.g. if user is using AWS EFA instead of InfiniBand).

Reply:
Synced offline, I will add:
1. V100 and H100 heuristics
2. user-facing configs for intra-node and inter-node interconnect bandwidth